File: Codeforces/Educational Codeforces Round 89/C-palindromicPaths.py
# auth: starkizard
# refer https://www.geeksforgeeks.org/zigzag-or-diagonal-traversal-of-matrix/ for diagonal order of a matrix
# okay let's say you are at index i,j arbitrary.
# you have path going down or right. let's say you end u on a and b in these two cases.
# now lets start from the mirror of it (cut diagonally) like if youre at 1 1, you are now at n m
# because we need to make all path palindrome the grid should be palindromic along the main diagonal
# now let's say this index is x,y and you need to go left or up, you should end up with the same a and b .
# and as we want to cover all the paths, down and right. we notice that the diagonals should be equal to their mirrors
# AND, they should only consist of 1s and 0s. 
# I tried hard explaining this approach of mine, But there's little i can do with comments only. A whiteboard would have been better xD

import os
import sys
import io
import math
def diagonalOrder(matrix,ROW,COL) : 
      
    # There will be ROW+COL-1 lines in the output 
    s=[]
    for line in range(1, (ROW + COL)) : 
        # Get column index of the first element 
        # in this line of output. The index is 0 
        # for first ROW lines and line - ROW for 
        # remaining lines  
        start_col = max(0, line - ROW) 
  
        # Get count of elements in this line. 
        # The count of elements is equal to 
        # minimum of line number, COL-start_col and ROW  
        count = min(line, (COL - start_col), ROW) 
  
        # Print elements of this line
        l=[]  
        for j in range(0, count) : 
            l.append(matrix[min(ROW, line) - j - 1] [start_col + j]) 
  
        s.append(l)
    return s
 
input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline
for _ in range(int(input())):
    n,m=map(int,input().split())
    matrix=[]
    for i in range(n):
        matrix.append(list(map(int,input().split())))
    s=diagonalOrder(matrix,n,m)
    if(len(s)%2==1):
        a=s[:len(s)//2]
        b=s[len(s)//2+1:][::-1]
    else:
        a=s[:len(s)//2]
        b=s[len(s)//2:][::-1]
    answer=0
    for i,j in zip(a,b):
        answer+=min(abs(i.count(0)+j.count(0)),abs(i.count(1)+j.count(1)))
    print(answer)

File: Codeforces/Educational Codeforces Round 89/C-palindromicPaths.cpp
//# refer https://www.geeksforgeeks.org/zigzag-or-diagonal-traversal-of-matrix/ for diagonal order of a matrix
//# Example: Rolling two dices would give us 36 pairs: (1,1),(1,2)...(6,6). Now visualize diagonal elements as pair having sum 2, pairs having sum 3 and so on.
//# Therefore the diagonal elements of matrix are pairs having same sum.
//# Given the word palindromic but what does it mean?
//# Palindromic means that elements from (1,1) and (n,m) are same 1 or 0.
//# When going from (1,1) we can go to either (1,2) or (2,1) so visualize this as diagonal with sum 3.
//# At the same time we can go from (n,m) to (n-1,m) and (n,m-1)
//# So question asks us to grab number of ways to make equidistant diagonal elements (from begin and end) equal.
//# We have taken k as the maximum limit of diagonal sum.
//# Run loop and see if i+j is p or is it the diagonal from the end (n+m-p)
//# Add the minimum of zeroes or one's
//# print the sum.

#include <bits/stdc++.h>

#define lli long long int
#define mod 1000000007
#define fi first
#define se second
#define ii pair<lli, lli>
#define mp make_pair
#define endl '\n'
const double PI = 3.141592653589793238460;
lli gcd(lli a, lli b) { if (a == 0) return b; return gcd(b % a, a);}
lli max(lli a, lli b) {if (a > b) return a; else return b;}
lli min(lli a, lli b) {if (a < b) return a; else return b;}
lli power(lli a,lli b) {lli i,p=1; for(i=0;i<b;i++){p*=a;} return p;}

using namespace std;

int main()
{
    ios::sync_with_stdio(false);

	lli t;
	
	cin>>t;
	
	while(t--)
	{
		
		lli n,m,i,j,k,p,s=0,r=0,c=0;
		
		cin>>n>>m;
		
		lli a[n][m];
		
		k = m + n-1;
		
		for(i=0;i<n;i++)
		{
			
			for(j=0;j<m;j++)
			{
				
				cin>>a[i][j];
				
			}
		}
		
		if(k%2==0)
		{
			
			k = k/2;
			
		}
		else
		{
			
			k = (k+1)/2;
			
			k--;
			
		}
		
		for(p=2;p<=k+1;p++)
		{
			s=0;
			
			c=0;
			
			for(i=0;i<n;i++)
			{
				
				for(j=0;j<m;j++)
				{
					
					if((i+j+2==p || i+j+2==n+m-p+2 ))
					{
						
						s += a[i][j];
						
						c++;
						
					}
					
				}
				
			}
			
			r += min(s,c-s);
			
		}
		
		cout<<r<<endl;
		
	}

    return 0;
}

